### PR TITLE
Add a new npx task to execute npm CLI tools thanks to the npx command

### PR DESCRIPTION
**Pull request by bsautel**
I created a new task type called NpxTask that is able to execute the NPM's `npx` command.

I had an issue in one of the integration test. I wanted to execute the `chcase-cli` NPM package and make it camel case a string. I expected the command result to be in the build output but it is not. I have many things (NPM logs for instance) but not that. It appears that all the logs I see come from the error output, whereas the command result is printed in the standard output which does not seem to be in the build output. As a workaround, I tested the with `--help` which prints the help in the error output, so I could get it in the build output.

It's probably normal that the standard output is not in the build output, but I could not find where it is configured. Any idea?

Hope it will be add some value to the plugin!

Don't hesitate to send me feedback to improve my code if it is not enough good. `NpmTask` and `NpxTasks` have some duplicated lines, it probably would be nice to remove this duplication. I can do it but I would like to ensure the rest of my contribution is right before doing that.

Note: I tested this version in one of my Angular projects which is built by Gradle and it works as expected.

### CODE REVIEW

**Comment by bsautel** [comment]
I added some integration tests to the NodeTask, NpmTask and NpxTask. They ensure the options are correctly taken into account. They also check that they declare correctly their input / output to ensure the up-to-date detection works as expected. I had to improve the inputs declarations of these tasks to get these tests pass.

This makes a big pull request, sorry for that, but all these commits are related (mainly because they share some test resources).

File: src/main/groovy/com/moowork/gradle/node/npm/NpmInstallTask.groovy [hunk 1]
```diff
@@ -59,7 +59,6 @@ class NpmInstallTask
 
     @OutputFile
     @Optional
-    @PathSensitive(PathSensitivity.RELATIVE)
```
Comment (deepy) [inline src/main/groovy/com/moowork/gradle/node/npm/NpmInstallTask.groovy]: Shouldn't this continue to be relative PathSensitive?

Comment (bsautel) [reply src/main/groovy/com/moowork/gradle/node/npm/NpmInstallTask.groovy]: I was reported as a warning in the [plugin build](https://travis-ci.org/node-gradle/gradle-node-plugin/builds/578305803#L253):

    Warning: Type 'com.moowork.gradle.node.npm.NpmInstallTask': property 'packageLockFileAsOutput' is annotated with @PathSensitive that is not allowed for @OutputFile properties.

Comment (bsautel) [reply src/main/groovy/com/moowork/gradle/node/npm/NpmInstallTask.groovy]: The [Gradle documentation](https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:up_to_date_checks) tells that it is only for inputs:
> Used with any input file property to tell Gradle to only consider the given part of the file paths as important. For example, if a property is annotated with `@PathSensitive(PathSensitivity.NAME_ONLY)`, then moving the files around without changing their contents will not make the task out-of-date.

Comment (deepy) [reply src/main/groovy/com/moowork/gradle/node/npm/NpmInstallTask.groovy]: Oh right, I guess that's just for inputs

**Comment by deepy** [comment]
Sorry about the slow feedback, I've been sick all week, but now that I've had a chance to look at it I can say that this looks great! Thank you for the amazing work!
The path-related tests fail on macOS (due to /var/folders/bl/ vs /private/var/folders/bl) but we'll work around that after getting this merged, I just gotta walk my dog and then I'll run a final test before merging

**Comment by deepy** [comment]
Tested it on a local project and everything seems fine! (except that I had misconfigured my local project, used `args` instead of `npmCommand`, which strangely worked, I guess we'll need to mention this in the release notes of 2.1.0)

**Comment by bsautel** [comment]
Sorry for the broken test on macOS, thanks for the fix. A big thanks for the review and for accepting my pull request. And no problem for the slow feedback, take care of your health!

In your last comment, you are talking about the `NpmTask`, right? If it is the case, it's normal. The command executed by the plugin is `npm` + `npmCommand` + `args`. You can use either `npmCommand` or `args`, it will work. If you use both, it's important to take care of the order.

I add that it would be nice to add the same kind of tests for the `YarnTask`. Its inputs are not correctly declared.

I removed multiple warnings regarding missing or wrong task inputs declarations. It remains two of them for each task:

1. One that tells that a `File` is annotated with `@Input` and should not. It sounds like it works as expected. Moreover, the `@Input` documentation seems to indicate that this is legal:
> This will cause the task to be considered out-of-date when the property has changed. When used on a `java.io.File` object that refers to a file or directory, the up-to-date check is only dependent on the path and not the contents of the file or directory. To make it depend on the contents, use `InputFile` or `InputDirectory` respectively.
2. Another warning indicating that the `runner.execOverrides` is not annotated with any input or output annotation. In fact, it is an input and it should be annotated with `@Input`. I tried to do it, but it sounds like as it is a closure, Gradle first executes it and checks whether the result changed or not since last execution. This closure takes a parameter (an `ExecSpec`) but Gradle does not provide it, so the execution throws an error which breaks the build. I did not find any solution. To remove this warning, we could annotate it with `@Internal`, but it is wrong. If the `execOverrides` closure changes, the task result is likely to be different and for this reason the task should not be considered as up-to-date. I did not find any simple solution. The only solution I have is to introduce an exec configuration object (a data structure) shared by all tasks that would contain all the options that are relevant. All these properties could be declared as `@Input`s. The `ExecSpec` would be configured by the plugin based on this data structure. This should work but this breaks backward compatibility since this would remove the `execOverrides` support. What do you think about that?

**Comment by deepy** [comment]
The problem is that the NpmTask I had used *only* args and had no npmCommand set, so maybe we should mark both as internal and have an "actual" one that consists of npmCommand + args, I'll hopefully be able to have a look at and give a response to 2. that this weekend.

I guess we should create some issues for these things, I'll also have a look at that :-)

**Comment by deepy** [comment]
So with just a minor part of the cold left I've revisited this:
args and npmCommand have been marked as Optional in the current snapshot, I guess in reality you want at least one of them to be set, but there's technically nothing wrong with setting neither.
I've also extracted parts of 1. into a an issue and I'm wondering if using Action<ExecSpec> might solve 2, I'll give that a try tomorrow and in a worst-case scenario I'll push a (disabled) failing test.

I'll also start summing up the changes into release notes since I can't think of anything blocking a 2.1.0 release

**Comment by bsautel** [comment]
You are right, we can use either `npmCommand` or `args`. We can also use both of them and none of them (it will show the help).

I noticed in the code two differences between these two properties:

- They have a different type (`String[]` and `List<?>`).
- They don't have the same default value. By default `npmCommand` is `null` whereas `args` is `[]`.

Since they tend to represent the same thing, are those differences relevant?

As far as I know, `@Optional` tells that Gradle has to accept the value not to be defined (`null`). If it's right, `@Optional` for `args` is not really necessary since it cannot be `null` (not `null` by default and setting it to `null` would produce a `NullPointerException`) .

Thanks for the issue for 1.

For 2., I believe you are right, an `Action` would probably solve the issue. But I am not sure in this case it would be still usable in the Groovy DSL as a closure (`{}`).

And that's great if a 2.1.0 will be soon released! Very nice!

**Comment by deepy** [comment]
I guess `@Optional` on `args` is mostly a hint for people looking at the code rather than Gradle.
As for types, I guess it's legacy and just there for simplicity, since `npmCommand` can only be `String[]` there's no need to do any casts and `args` passed can be various types and will be handled by `ExecSpec`, but beyond that I can't think of any good reason to keep them separate

I've summed up the changes in 2.1.0 and it's quite a large one, so I'll do some more testing and then release it soon.
#34 can wait a version and so can execOverride, though from the brief testing I did passing a closure to something wanting an Action worked (although if you're statically typed the syntax became awkward).
Once again, thank you for the amazing work! :-)

**Comment by bsautel** [comment]
`@Optional` is more than a simple hint. The build will be broken if a non `@Optional` `@Input` property is set to `null`. It's the same for `@InputFile`, if the `package-lock.json` was not `@Optional` in the `NpmInstall` task, Gradle would not want to build when the file does not exists (whereas it's optional in npm).

To sum-up, you mean that the `args` field is only here for backward-compatibility, right?

And in this case, shouldn't we have the same approach in the new `NpxTask` before releasing it? I created two parameters, one for the command and the other one for the arguments, but we maybe should merge them in the command one which should be an array and no longer a list. This would be consistent with the `NpmTask`. I can make a pull request to change that if you want.

And good news for the `Action`!

Happy to contribute to this tool!

**Comment by deepy** [comment]
`@InputFile` is actually pretty sneaky, if you have a `File` with `.exists() == false` it'll throw an error even if it's marked as `@Optional`.
In the case of `NpmInstall`s `args` the `@Optional` the idea was to retain the same error message, though you are right in that it'll throw a NPE eventually if someone sets it to null, and when I added it I was mostly thinking of how bad the Gradle message for an `@Optional @InputFile` which `.exist() == false` is.

I guess there really isn't any downside to merging them into a single list, except for breaking backwards wombat, we can add arguments by adding them at the start or end of the list.
Well, maybe if we or users provide default arguments that are supposed to be replaced.

In the case of `NpxTask` I guess keeping it as it is will be fine, since it's new there's no previous users and on account of the typing `npmCommand` should probably be preferred over `args`.

**Comment by bsautel** [comment]
I did not know about this specificity of `@Optional` and `@InputFile`. I would not expect that, but there may be a good reason. I now understand why there is this code for instance in the `NpmInstall` task inputs declaration:
    return file.exists() ? file : null

Thanks for all your answers!

**Comment by deepy** [comment]
If I remember correctly this changed between Gradle 4 and Gradle 5, in Gradle 5 it turned into an error and emerged painfully in version 1.5.0 https://github.com/node-gradle/gradle-node-plugin/blob/master/CHANGELOG.md#version-151-2019-06-19

**Pull request merged**